File: 17 Letter Combinations of a Phone Number/17 Letter Combinations of a Phone Number.go
func letterCombinations(digits string) []string {
    var cache = map[int]string{
        2:"abc",
        3:"def",
        4:"ghi",
        5:"jkl",
        6:"mno",
        7:"pqrs",
        8:"tuv",
        9:"wxyz",
    }
    var ans = []string{}
    var getComb func(string, string)
    getComb = func(selected string, rest string){
        currentDigit, _ := strconv.Atoi(string(rest[0]))
        for _, ch := range cache[currentDigit] {
            if len(rest[1:]) == 0 {
                ans = append(ans, selected + string(ch))
            } else {
                getComb(selected+string(ch), rest[1:])
            }
        }
    }
    if len(digits)>0 {
        getComb("", digits)
    }
    return ans
}

File: 17 Letter Combinations of a Phone Number/17 Letter Combinations of a Phone Number.py
"""
Given a string containing digits from 2-9 inclusive, return all possible letter combinations that the number could represent. Return the answer in any order.
A mapping of digit to letters (just like on the telephone buttons) is given below. Note that 1 does not map to any letters.

Example 1:

Input: digits = "23"
Output: ["ad","ae","af","bd","be","bf","cd","ce","cf"]

Example 2:

Input: digits = ""
Output: []

Example 3:

Input: digits = "2"
Output: ["a","b","c"]
"""
class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        interpret_digit = {
            '1': '',
            '2': 'abc',
            '3': 'def',
            '4': 'ghi',
            '5': 'jkl',
            '6': 'mno',
            '7': 'pqrs',
            '8': 'tuv',
            '9': 'wxyz',
            '0': ' '}

        all_combinations  = [''] if digits else []

        for digit in digits :
            curr_combination = list()
            for letter in interpret_digit[digit]:
                for combination in all_combinations:
                    curr_combination.append(combination+letter)

            all_combinations = curr_combination
        return all_combinations
